core/block_validator.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2014 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/pow"
+	"gopkg.in/fatih/set.v0"
+)
+
+type BlockValidator struct {
+	// Canonical block chain
+	bc *BlockChain
+	// Proof of work used for validating
+	Pow pow.PoW
+}
+
+// NewBlockValidator returns a new block validator which is safe for re-use
+func NewBlockValidator(blockchain *BlockChain, pow pow.PoW) *BlockValidator {
+	sm := &BlockValidator{
+		Pow: pow,
+		bc:  blockchain,
+	}
+	return sm
+}
+
+func (v *BlockValidator) ValidateBlock(block *types.Block) error {
+	if v.bc.HasBlock(block.Hash()) {
+		return &KnownBlockError{block.Number(), block.Hash()}
+	}
+	if !v.bc.HasBlock(block.ParentHash()) {
+		return ParentError(block.ParentHash())
+	}
+	var (
+		parent = v.bc.GetBlock(block.ParentHash())
```

Comment:
When we check the parent's existence above, we already retrieve it from the database. I think it would be a bit more optimal to simply retrieve it using `GetBlock` above instead of `HasBlock` and check it against nil. Even though the cache in the blockchain should short circuit the extra query, I really don't see the need for a pre-check if we're going to pull it anyway immediately afterwards.